README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# visualizeR <img src="man/figures/logo.png" align="right" alt="" width="120"/>

> What a color\! What a viz\!

`visualizeR` proposes some utils to get REACH and AGORA colors,
ready-to-go color palettes, and a few visualization functions
(horizontal hist graph for instance).

## Installation

You can install the last version of visualizeR from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("gnoblet/visualizeR", build_vignettes = TRUE)
```

## Roadmap

Roadmap is as follows:

  - \[X\] Add IMPACT’s colors
  - \[X\] Add all color palettes from the internal documentation
  - \[ \] There remains to be added more-than-7-color palettes and black
    color palettes
  - \[X\] Add new types of visualization (e.g. dumbbell plot, lollipop
    plot, etc.)
  - \[X\] Use examples
  - \[ \] Add some ease-map functions
  - \[ \] Add some interactive functions (maps and graphs)
  - \[ \] Consolidate and make errors transparent

## Request

Please, do not hesitate to pull request any new viz or colors or color
palettes, or to email request any change
(<guillaume.noblet@reach-initiative.org> or <gnoblet@zaclys.net>).

## Colors

Color palettes for REACH, AGORA and IMPACT are available. Functions to
access colors and palettes are `cols_initiative()` or
`pal_initiative()`. For now, the initiative with the most colors and
color palettes is REACH. Feel free to pull requests new AGORA and IMPACT
colors.

``` r
library(visualizeR)

# Get all saved REACH colors, named
cols_reach(unnamed = F)[1:10]
#>        white        black    main_grey     main_red main_lt_grey   main_beige 
#>    "#FFFFFF"    "#000000"    "#58585A"    "#EE5859"    "#C7C8CA"    "#D2CBB8" 
#>     iroise_1     iroise_2     iroise_3     iroise_4 
#>    "#DFECEF"    "#B1D7E0"    "#699DA3"    "#236A7A"

# Extract a color palette as hexadecimal codes and reversed
pal_reach(palette = "main", reversed = TRUE, color_ramp_palette = FALSE)
#> [1] "#58585A" "#EE5859" "#C7C8CA" "#D2CBB8"

# Get all color palettes names
pal_reach(show_palettes = T)
#>  [1] "main"            "primary"         "secondary"       "two_dots"       
#>  [5] "two_dots_flashy" "red_main"        "red_main_5"      "red_alt"        
#>  [9] "red_alt_5"       "iroise"          "iroise_5"        "discrete_6"     
#> [13] "red_2"           "red_3"           "red_4"           "red_5"          
#> [17] "red_6"           "red_7"           "green_2"         "green_3"        
#> [21] "green_4"         "green_5"         "green_6"         "green_7"        
#> [25] "artichoke_2"     "artichoke_3"     "artichoke_4"     "artichoke_5"    
#> [29] "artichoke_6"     "artichoke_7"     "blue_2"          "blue_3"         
#> [33] "blue_4"          "blue_5"          "blue_6"          "blue_7"
```

## Charts

### Example 1: Bar chart, already REACH themed

``` r
library(visualizeR)
library(palmerpenguins)
library(dplyr)

df <- penguins |> 
  group_by(island, species) |> 
  summarize(
    mean_bl = mean(bill_length_mm, na.rm = T),
    mean_fl = mean(flipper_length_mm, na.rm = T)) |> 
  ungroup()

# Simple bar chart by group with some alpha transparency
bar(df, island, mean_bl, species, percent = FALSE, alpha = 0.6, x_title = "Mean of bill length")
```

<img src="man/figures/README-example-bar-chart-1.png" width="65%" />

``` r

# Using another color palette through `theme_reach()` and changing scale to percent
bar(df, island,mean_bl, species, percent = TRUE, theme = theme_reach(palette = "artichoke_3"))
```

<img src="man/figures/README-example-bar-chart-2.png" width="65%" />

``` r

# Not flipped, with text added, group_title, no y-axis and no bold for legend
bar(df, island, mean_bl, species, group_title = "Species", flip = FALSE, add_text = TRUE, add_text_suffix = "%", percent = FALSE, theme = theme_reach(text_font_face = "plain", axis_y = FALSE))
```

<img src="man/figures/README-example-bar-chart-3.png" width="65%" />

### Example 2: Point chart, already REACH themed

At this stage, `point_reach()` only supports categorical grouping colors
with the `group` arg.

``` r

# Simple point chart
point(penguins, bill_length_mm, flipper_length_mm)
```

<img src="man/figures/README-example-point-chart-1.png" width="65%" />

``` r

# Point chart with grouping colors, greater dot size, some transparency, reversed color palette
point(penguins, bill_length_mm, flipper_length_mm, island, alpha = 0.6, size = 3, theme = theme_reach(reverse = TRUE))
```

<img src="man/figures/README-example-point-chart-2.png" width="65%" />

``` r

# Using another color palettes
point(penguins, bill_length_mm, flipper_length_mm, island, size = 1.5, x_title = "Bill", y_title = "Flipper", title = "Length (mm)", theme = theme_reach(palette = "artichoke_3", text_font_face = , grid_major_x = TRUE,  title_position_to_plot = FALSE))
```

<img src="man/figures/README-example-point-chart-3.png" width="65%" />

### Example 3: Dumbbell plot, REACH themed

Remember to ensure that your data are in the long format and you only
have two groups on the x-axis; for instance, IDP and returnee and no NA
values.

``` r
# Prepare long data
df <- tibble::tibble(
  admin1 = rep(letters[1:8], 2),
  setting = c(rep(c("Rural", "Urban"), 4), rep(c("Urban", "Rural"), 4)),
  stat = rnorm(16, mean = 50, sd = 18)
) |>
  dplyr::mutate(stat = round(stat, 0))

# Example, adding a parameter to `theme_reach()` passed on `ggplot2::theme()` to align legend title

dumbbell(df,
         stat,
         setting,
         admin1,
         title = "% of HHs that reported open defecation as sanitation facility",
         group_y_title = "Admin 1",
         group_x_title = "Setting",
         theme = theme_reach(legend_position =  "bottom",
                             legend_direction = "horizontal",
                             legend_title_font_face = "bold",
                             palette = "primary",
                             title_position_to_plot = FALSE,
                             legend.title.align = 0.5)) +
  # Change legend title position (could be included as part of the function)
  ggplot2::guides(  
    color = ggplot2::guide_legend(title.position = "left"),
    fill =  ggplot2::guide_legend(title.position = "left")
  )
```

<img src="man/figures/README-example-dumbbell-plot-1.png" width="65%" />

### Example 4: donut chart, REACH themed (to used once, not twice)

``` r

# Some summarized data: % of HHs by displacement status
df <- tibble::tibble(
  status = c("Displaced", "Non displaced", "Returnee", "Don't know/Prefer not to say"),
  percentage = c(18, 65, 12, 3)
)

# Donut
donut(df, 
      status, 
      percentage, 
      hole_size  = 3, 
      add_text_suffix = "%", 
      add_text_color = cols_reach("dk_grey"), 
      add_text_treshold_display = 5,
      x_title = "Displacement status", 
      title = "% of HHs by displacement status", 
      theme = theme_reach(legend_reverse = TRUE))
```

<img src="man/figures/README-example-donut-plot-1.png" width="65%" />

### Example 5: waffle chart

``` r
#
waffle(df, status, percentage, x_title = "A caption", title = "A title", subtitle = "A subtitle")
```

<img src="man/figures/README-example-waffle-plot-1.png" width="65%" />

### Example 6: alluvial chart, REACH themed

``` r

# Some summarized data: % of HHs by self-reported status of displacement in 2021 and in 2022
df <- tibble::tibble(
  status_from = c(rep("Displaced", 4),
                  rep("Non displaced", 4),
                  rep("Returnee", 4),
                  rep("Dnk/Pnts", 4)),
  status_to = c("Displaced", "Non displaced", "Returnee", "Dnk/Pnts", "Displaced", "Non displaced", "Returnee", "Dnk/Pnts", "Displaced", "Non displaced", "Returnee", "Dnk/Pnts", "Displaced", "Non displaced", "Returnee", "Dnk/Pnts"),
  percentage = c(20, 8, 18, 1, 12, 21, 0, 2, 0, 3, 12, 1, 0, 0, 1, 1)
)

# Alluvial, here the group is the status for 2021

alluvial(df, 
         status_from, 
         status_to,
         percentage, 
         status_from,
         from_levels = c("Displaced", "Non displaced", "Returnee", "Dnk/Pnts"), 
         alpha = 0.8, 
         group_title = "Status for 2021",
         title = "% of HHs by self-reported status from 2021 to 2022", 
         theme = theme_reach(
           axis_y = FALSE, 
           legend_position = "none"))
```

<img src="man/figures/README-example-alluvial-plot-1.png" width="65%" />

### Example 7: lollipop chart

``` r
library(tidyr)
# Prepare long data
df <- tibble::tibble(
 admin1 = replicate(15, sample(letters, 8)) |> t() |> as.data.frame() |> unite("admin1", sep = "") |> dplyr::pull(admin1),  
 stat = rnorm(15, mean = 50, sd = 15)) |>
  dplyr::mutate(stat = round(stat, 0))

# Make lollipop plot, REACH themed, vertical with 45 degrees angle X-labels
lollipop(df,
         admin1,
         stat,
         arrange = FALSE,
         add_text = FALSE,
         flip = FALSE,
         y_title = "% of HHs",
         x_title = "Admin 1",
         title = "% of HHs that reported having received a humanitarian assistance",
         theme = theme_reach(axis_text_x_angle = 45, 
                             grid_major_y = TRUE, 
                             grid_major_y_size = 0.2, 
                             grid_major_x = TRUE, 
                             grid_minor_y = TRUE))
```

<img src="man/figures/README-example-lollipop-chart-1.png" width="65%" />

``` r

# Horizontal, greater point size, arranged by value, no grid, and text labels added
lollipop(df,
         admin1,
         stat,
         arrange = TRUE,
         point_size = 10,
         point_color = cols_reach("main_beige"),
         segment_size = 2,
         add_text = TRUE,
         add_text_suffix = "%",
         y_title = "% of HHs",
         x_title = "Admin 1",
         title = "% of HHs that reported having received a humanitarian assistance in the 12 months prior to the assessment",
         theme = theme_reach(title_position_to_plot = FALSE))
```

<img src="man/figures/README-example-lollipop-chart-2.png" width="65%" />

## Maps

``` r

# Add indicator layer 
# - based on "pretty" classes and title "Proportion (%)" 
# - buffer to add a 10% around the bounding box
map <- add_indicator_layer(
  indicator_admin1, 
  opn_dfc,
  buffer = 0.1) + 
  # Layout - some defaults - add the map title
  add_layout("% of HH that reported open defecation as sanitation facility") + 
  # Admin boundaries as list of shape files (lines) and colors, line widths and labels as vectors
  add_admin_boundaries(
    lines = list(line_admin1, border_admin0, frontier_admin0),
    colors = cols_reach("main_lt_grey", "dk_grey", "black"),
    lwds = c(0.5, 2, 3),
    labels = c("Department", "Country", "Dominican Rep. frontier"),
    title = "Administrative boundaries") + 
  # Add text labels - centered on admin 1 centroids
  add_admin_labels(centroid_admin1, ADM1_FR_UPPER) +
  # Add a compass
  add_compass() +
  # Add a scale bar
  add_scale_bar() +
  # Add credits
  add_credits("Admin. boundaries. : CNIGS \nCoord. system: GCS WGS 1984")
```

![Once exported with
`tmap::tmap_save()`.](man/figures/README-example-map.png)


NEWS.md
# visualizeR 0.8.9000

* Add `waffle()`.
* Add a `pal_default()` function as a fallback for scale functions and as an initiative in `theme_reach()`. It uses `viridis::magma()`.
* Update of AGORA palettes.
* Some other small bug fixes.

---

# visualizeR 0.7.9000

* Add `dumbbell()`.
* Add `alluvial()`
* Add `donut()`
* Add `lollipop()`
* Add further parameters to `theme_reach()`, including grid lines args.

---

# visualizeR 0.6.9000

* Add `dumbbell()`.
* Add further parameters to `theme_reach()`.

---

# visualizeR 0.5.9000

* Add wrapping of title, subtitle and caption thanks to `ggtext`
* Add wrapping of labels for `bar()` x-discrete scale.
* Add further parameters to `theme_reach()`

---

# visualizeR 0.4.9000

* Breaking changes: remove dependency to `ggblanket`.
* Full rewrite of `theme_reach()`.
* `bar_reach` is now `bar()` and theming is passed through argument `theme` for which default is `theme_reach()`.
* `point_reach` is now `point()` and theming is passed through argument `theme` for which default is `theme_reach()`.

---

# visualizeR 0.3.9000

* Breaking changes: update to `ggblanket` v1.6.1.
* Add plotting functions for indicator maps.

---

# visualizeR 0.2.9000

* Breaking changes: almost all functions got refinements, and there are new functions, typically `hbar()` becomes `bar_reach()` and `point_reach()` is added.
* Following `theme_reach()` is now used by all plotting functions.
* Add README.md.

---

# visualizeR 0.1.7.9000

* Fixed some color palettes.

---

# visualizeR 0.1.6.9000

* IMPACT colors and palettes are added: function `cols_impact()` `pal_impact()`.
* Color palettes from REACH are added (2 to 7 continuous palettes) ; see updated `cols_reach()` and `pal_reach()`.

---

# visualizeR 0.1.5.9000

* Move from `simplevis` to successor `ggblanket`.

---

# visualizeR 0.1.4.9000

* `hbar()` gains a new boolean argument `reverse` to pass to `pal_reach()` or `pal_agora()`, indicating if the color palette should be reversed or not.

---

# visualizeR 0.1.3.9000

* Small change to `hbar()`: removes error arg within `simplevis::gg_hbar()` call.

---

# visualizeR 0.1.2.9000

* There was a duplicate `scale_color()` function, which should have been and is now `scale_fill()`

--- 

# visualizeR 0.1.1.9000

* Added two horizontal bar functions: `hbar()`, `hbar_percent()` (#3)
* Added some internals to check for missing columns and bad arguments (#3)
* Modified some `theme_reach()` documentation
* Add `buffer_bbox()` function to produce a buffered bbox, e.g. for use with `tmap`

--- 

# visualizeR 0.1.0

* Added a `NEWS.md` file to track changes to the package
* Initiate repo


R/map.R


#' Wrapper around `tmap::tm_polygons()` with sane defaults for plotting indicator values
#'
#' @param poly Multipolygon shape defined by sf package.
#' @param col Numeric attribute to map.
#' @param buffer A buffer, either one value or a vector of 4 values (left, bottom, right, top).
#' @param n The desire number of classes.
#' @param style Method to process the color scale for continuous numerical variables. See `classInt::classIntervals()` for details.
#' @param palette Vector of fill colors as hexadecimal values. For REACH color palettes, it is possible to use `pal_reach()`. For now,'palette' must be changed manually, accordingly to the number of drawn classes.
#' @param as_count Boolean. When col is a numeric variable, should it be processed as a count variable? For instance, 0, 1-10, 11-20.
#' @param color_na Fill color for missing data.
#' @param text_na Legend text for missing data.
#' @param legend_title Legend title.
#' @param legend_text_separator Text separator for classes. E.g. " to " will give 0, 1 to 10, 11 to 20.
#' @param border_alpha Transparency of the border.
#' @param border_col Color of the border.
#' @param lwd Linewidth of the border.
#' @param ... Other arguments to pass to `tmap::tm_polygons()`.
#'
#' @return A tmap layer.
#' @export
#'
add_indicator_layer <- function(
    poly,
    col,
    buffer = NULL,
    n = 5,
    style = "pretty",
    palette = pal_reach("red_5"),
    as_count = TRUE,
    color_na = cols_reach("white"),
    text_na = "Missing data",
    legend_title = "Proportion (%)",
    legend_text_separator = " - ",
    border_alpha = 1,
    border_col = cols_reach("lt_grey_1"),
    lwd = 1,
    ...){

  #------ Checks and make valid

  rlang::check_installed("tmap", reason = "Package \"tmap\" needed for `add_indicator_layer()` to work. Please install it.")

  poly <- sf::st_make_valid(poly)

  #------ Other checks

  col_name <- rlang::as_name(rlang::enquo(col))
  if_not_in_stop(poly, col_name, "poly", "col")

  if (!is.numeric(poly[[col_name]])) rlang::abort(glue::glue("{col_name} is not numeric."))


  #------ Prepare data

  if(!is.null(buffer)){ buffer <- buffer_bbox(poly, buffer) } else { buffer <- NULL }


  #------ Polygon layer

  layer <- tmap::tm_shape(
    poly,
    bbox = buffer
  ) +
    tmap::tm_polygons(
      col = col_name,
      n = n,
      style = style,
      palette = palette,
      as.count = as_count,
      colorNA = color_na,
      textNA = text_na,
      title = legend_title,
      legend.format = list(text.separator = legend_text_separator),
      borderl.col = border_col,
      border.alpha = border_alpha,
      lwd = lwd,
      ...
    )

  return(layer)

}




#' Add admin boundaries (lines) and the legend
#'
#' @param lines List of multiline shape defined by sf package.
#' @param colors Vector of hexadecimal codes. Same order as lines.
#' @param labels Vector of labels in the legend. Same order as lines.
#' @param lwds Vector of line widths. Same order as lines.
#' @param title Legend title.
#' @param buffer A buffer, either one value or a vector of 4 values (left, bottom, right, top).
#' @param ... Other arguments to pass to each shape in `tmap::tm_lines()`.
#'
#' @return A tmap layer.
#' @export
#'
add_admin_boundaries <- function(lines, colors, labels, lwds, title = "", buffer = NULL, ...){


  #------ Package check

  rlang::check_installed("tmap", reason = "Package \"tmap\" needed for `add_admin_boundaries()` to work. Please install it.")


  #------ Check that the length of vectors is identical between arguments

  if(!inherits(lines, "list")) rlang::abort("Please provide a list for lines.")

  ll <- list(lines, colors, labels, lwds)
  if (!all(sapply(ll,length) == length(ll[[1]]))) rlang::abort("lines, colors, labels, lwds do not all have the same length.")


  #------ Make valid

  lines <- lapply(lines, \(x) sf::st_make_valid(x))


  #------ Prepare legend
  legend_lines <- tmap::tm_add_legend("line",
                                      title = title,
                                      col = colors,
                                      lwd = lwds,
                                      labels = labels)


  #------ Let's go with all line shapes

  if(!is.null(buffer)){ buffer <- buffer_bbox(lines[[1]], buffer) } else { buffer <- NULL }


  layers <- tmap::tm_shape(lines[[1]], bbox = buffer) +
    tmap::tm_lines(lwd = lwds[[1]], col = colors[[1]], ...)

  if (length(lines) == 1) {

    layers <- layers + legend_lines

    return(layers)

  } else {

    for(i in 2:length(lines)){

      layers <- layers + tmap::tm_shape(shp = lines[[i]]) + tmap::tm_lines(lwd = lwds[[i]], col = colors[[i]], ...)
    }

    layers <- layers + legend_lines

    return(layers)

  }
}




#' Basic defaults based on `tmap::tm_layout()`
#'
#' @param title Map title.
#' @param legend_position Legend position. Not above the map is a good start.
#' @param frame Boolean. Legend frame?
#' @param legend_frame Legend frame color.
#' @param legend_text_size Legend text size in 'pt'.
#' @param legend_title_size Legend title size in 'pt'.
#' @param title_size Title text size in 'pt'.
#' @param title_fontface Title fontface. Bold if you wanna exemplify a lot what it is about.
#' @param title_color Title font color.
#' @param fontfamily Overall fontfamily. Leelawadee is your precious.
#' @param ... Other arguments to pass to `tmap::tm_layout()`.
#'
#' @return A tmap layer.
#' @export
#'
add_layout <- function(
    title = NULL,
    legend_position = c(0.02, 0.5),
    frame = FALSE,
    legend_frame = cols_reach("main_grey"),
    legend_text_size = 0.6,
    legend_title_size = 0.8,
    title_size = 0.9,
    title_fontface = "bold",
    title_color = cols_reach("main_grey"),
    # check.and.fix = TRUE,
    fontfamily = "Leelawadee",
    ...){

  layout <- tmap::tm_layout(
    title = title,
    legend.position = legend_position,
    legend.frame = legend_frame,
    frame = FALSE,
    legend.text.size = legend_text_size,
    legend.title.size = legend_title_size,
    title.size = title_size,
    title.fontface = title_fontface,
    title.color = title_color,
    fontfamily = fontfamily,
    ...)

  return(layout)

  }




#' Wrapper around `tmap::tm_text()` with sane defaults for plotting admin labels.
#'
#' @param point Multipoint shape defined by sf package.
#' @param text Text labels column.
#' @param size Relative size of the text labels.
#' @param fontface Fontface.
#' @param fontfamily Fontfamily. Leelawadee is your precious.
#' @param shadow Boolean. Add a shadow around text labels. Issue opened on Github to request.
#' @param auto_placement Logical that determines whether the labels are placed automatically.
#' @param remove_overlap Logical that determines whether the overlapping labels are removed.
#' @param ... Other arguments to pass to `tmap::tm_text()`.
#'
#' @return A tmap layer.
#' @export
#'
add_admin_labels <- function(point,
                             text,
                             size = 0.5,
                             fontface = "bold",
                             fontfamily = "Leelawadee",
                             shadow = TRUE,
                             auto_placement = FALSE,
                             remove_overlap = FALSE,
                             ...){


  #------ Restrictive sf checks (might not be necessary depending on the desired behaviour)

  rlang::check_installed("tmap", reason = "Package \"tmap\" needed for `add_indicator_layer()` to work. Please install it.")

  point <- sf::st_make_valid(point)

  #------ Other checks

  text_name <- rlang::as_name(rlang::enquo(text))
  if_not_in_stop(point, text_name, "point", "text")

  #------ Point text layer

  layer <- tmap::tm_shape(point) +
    tmap::tm_text(text = text_name,
                  size = size,
                  fontface = fontface,
                  fontfamily = fontfamily,
                  shadow = shadow,
                  auto.placement = auto_placement,
                  remove.overlap = remove_overlap,
                  ...)

  return(layer)

}




#' Add a compass
#'
#' @param text_size Relative font size.
#' @param position Position of the compass. Vector of two values, specifying the x and y coordinates.
#' @param color_dark Color of the dark parts of the compass.
#' @param text_color color of the text.
#' @param type Compass type, one of: "arrow", "4star", "8star", "radar", "rose".
#' @param ... Other arguments to pass to `tmap::tm_compass()`.
#'
#' @return A tmap layer.
#' @export
#'
add_compass <- function(text_size = 0.6,
                        position = c("right", 0.8),
                        color_dark = cols_reach("black"),
                        text_color = cols_reach("black"),
                        type = "4star",
                        ...){

  compass <- tmap::tm_compass(
    text.size = text_size,
    position = position,
    color.dark = color_dark,
    type = type,
    text.color = text_color
  )

  return(compass)

}




#' Add a scale bar
#'
#' @param text_size Relative font size.
#' @param position Position of the compass. Vector of two values, specifying the x and y coordinates.
#' @param color_dark Color of the dark parts of the compass.
#' @param breaks Breaks of the scale bar. If not specified, breaks will be automatically be chosen given the prefered width of the scale bar. Example: c(0, 50, 100).
#' @param ... Other arguments to pass to `tmap::tm_compass()`.
#'
#' @return A tmap layer.
#' @export
#'
add_scale_bar <- function(text_size = 0.6,
                          position = c("left", 0.01),
                          color_dark = cols_reach("black"),
                          breaks = c(0, 50, 100),
                          ...){

  scale_bar <- tmap::tm_scale_bar(
    text.size = text_size,
    position = position,
    color.dark = color_dark,
    breaks = breaks,
    ...
  )

  return(scale_bar)

}




#' Do you want to credit someone or some institution?
#'
#' @param text Text.
#' @param size Relative text size.
#' @param bg_color Background color.
#' @param position Position. Vector of two coordinates. Usually somewhere down.
#' @param ... Other arguments to pass to `tmap::tm_credits()`.
#'
#' @return A tmap layer.
#' @export
#'
add_credits <- function(text, size = 0.4, bg_color = NA, position = c(0.75, 0.02), ...){

  tmap::tm_credits(text,
                   size = size,
                   bg.color = bg_color,
                   position = position,
                   ...)
}

